### PR TITLE
bazel: fix false rebuilds of c-ares library.

### PR DESCRIPTION
**Pull request by htuch**
This patch undoes an evil hack where we were relying on Bazel internal
directory layout to cope with c-ares build limitations. It also no long
requires that sandboxing be disabled. It's still ugly, but less so.

### CODE REVIEW

**Comment by htuch** [comment]
@jamessynge @dnoe

File: bazel/repositories.bzl [hunk 1]
```diff
@@ -91,19 +91,28 @@ cc_library(
...
+    "./buildconf",
+    "./configure",
+    "popd",
+    "cp $(@D)/cares_build/external/cares_git/ares_config.h $@",
+]
+
 genrule(
     name = "config",
     # This list must not contain ares_config.h.
-    srcs = glob(["**/*.m4"]),
```
Comment (dnoe) [inline bazel/repositories.bzl]: This looks sneakily unrelated to the other changes, is it correct?

Comment (htuch) [inline bazel/repositories.bzl]: We need more files now that we're not illegally stealing them from an internal directory :) It doesn't have to be that tight a bound, this glob only hits everything in the c-ares source tree.

**Comment by htuch** [comment]
@lizan for Bazel expertise.

File: bazel/repositories.bzl [hunk 2]
```diff
@@ -91,19 +91,27 @@ cc_library(
     visibility = ["//visibility:public"],
 )
 
+# There is some moderate evil here. buildconf needs to be invoked in its source
+# location, where it also generates temp files (no option to place in TMPDIR).
+# Also, configure does a cd relative dance that doesn't play nice with the
+# symlink execroot in the Bazel build. So, copy everything prior to building.
+ARES_CONFIG_SH_CMDS = [
+    "export TMPDIR=$$(realpath $(@D))",
```
Comment (lizan) [inline bazel/repositories.bzl]: What is the reason to point TMPDIR to a path under `$(@D)` instead of real TMPDIR?

File: bazel/repositories.bzl [hunk 3]
```diff
@@ -91,19 +91,27 @@ cc_library(
...
+    "rm -rf $(@D)/cares_build",
+    "mkdir -p $(@D)/cares_build",
+    "rsync -aRLv $(SRCS) $(@D)/cares_build",
```
Comment (lizan) [inline bazel/repositories.bzl]: I would not assuming `rsync` available on the machine, can't this be done by `cp` with `-L`?

File: bazel/repositories.bzl [hunk 4]
```diff
@@ -91,19 +91,27 @@ cc_library(
...
     outs = ["ares_config.h"],
-    # There is some serious evil here. buildconf needs to be invoked in its source location, where
-    # it also generates temp files (no option to place in TMPDIR). Also, configure does a cd
-    # relative dance that doesn't play nice with the symlink execroot in the Bazel build. So,
-    # disable sandboxing and do some fragile stuff with the build dirs.
-    # TODO(htuch): Figure out a cleaner way to handle this.
-    cmd = "pushd ../../external/cares_git; ./buildconf; ./configure; " +
-          "cp -f ares_config.h ../../execroot/envoy/$@",
-    local = 1,
+    cmd = "; ".join(ARES_CONFIG_SH_CMDS),
```
Comment (lizan) [inline bazel/repositories.bzl]: `&&` instead of `;` ?

File: bazel/repositories.bzl (hunk 2 above)
Comment (htuch) [inline bazel/repositories.bzl]: This is needed as `TMPDIR` might not be `/tmp`, and sandboxing doesn't like it unless you stick to `@D` or `/tmp`.

**Pull request merged**